src/improvements/AddressRegistry.sol

```diff
@@ -181,4 +278,178 @@ contract AddressRegistry is IAddressRegistry, Test {
+        uint256 chainId,
+        string memory contractIdentifier
+    ) internal pure returns (address) {
+        return vm.parseJsonAddress(
+            chainAddressesContent, string.concat("$.", vm.toString(chainId), ".", contractIdentifier)
+        );
+    }
+
+    function getGuardian(address portal) internal view returns (address) {
+        try IFetcher(portal).guardian() returns (address guardian) {
+            return guardian;
+        } catch {
+            return IFetcher(portal).GUARDIAN();
+        }
+    }
+
+    function getSystemConfigProxy(address portal) internal view returns (address) {
+        try IFetcher(portal).systemConfig() returns (address systemConfig) {
+            return systemConfig;
+        } catch {
+            return IFetcher(portal).SYSTEM_CONFIG();
+        }
+    }
+
+    function getOptimismPortalProxy(address l1CrossDomainMessengerProxy) internal view returns (address) {
+        try IFetcher(l1CrossDomainMessengerProxy).PORTAL() returns (address optimismPortal) {
+            return optimismPortal;
+        } catch {
+            return IFetcher(l1CrossDomainMessengerProxy).portal();
+        }
+    }
+
+    function getAddressManager(address l1CrossDomainMessengerProxy) internal view returns (address addressManager) {
+        uint256 ADDRESS_MANAGER_MAPPING_SLOT = 1;
+        bytes32 slot = keccak256(abi.encode(l1CrossDomainMessengerProxy, ADDRESS_MANAGER_MAPPING_SLOT));
+        addressManager = address(uint160(uint256((vm.load(l1CrossDomainMessengerProxy, slot)))));
+    }
+
+    function getL1ERC721BridgeProxy(address systemConfigProxy, string memory chainAddressesContent, uint256 chainId)
+        internal
+        view
+        returns (address)
+    {
+        try IFetcher(systemConfigProxy).l1ERC721BridgeProxy() returns (address l1ERC721BridgeProxy) {
+            return l1ERC721BridgeProxy;
+        } catch {
+            return parseContractAddress(chainAddressesContent, chainId, "L1ERC721BridgeProxy");
+        }
+    }
+
+    function getOptimismMintableERC20FactoryProxy(
+        address systemConfigProxy,
+        string memory chainAddressesContent,
+        uint256 chainId
+    ) internal view returns (address) {
+        try IFetcher(systemConfigProxy).optimismMintableERC20Factory() returns (
+            address optimismMintableERC20FactoryProxy
+        ) {
+            return optimismMintableERC20FactoryProxy;
+        } catch {
+            return parseContractAddress(chainAddressesContent, chainId, "OptimismMintableERC20FactoryProxy");
+        }
+    }
+
+    function getDisputeGameFactoryProxy(address systemConfigProxy) internal view returns (address) {
+        try IFetcher(systemConfigProxy).disputeGameFactory() returns (address disputeGameFactoryProxy) {
+            return disputeGameFactoryProxy;
+        } catch {
+            return address(0); // Older chains don't have a dispute game factory and it should be discovered from the SystemConfig.
```

Comment:
nit, this code comment is confusing, since if they don't have a dispute game factory we can't discover it
```suggestion
            // Older chains don't have a dispute game factory, they have the L2OutputOracle
            return address(0);
```

Reply:
https://github.com/ethereum-optimism/superchain-ops/pull/565/commits/d8d4b196c692eac2c67a859d1bb3e5cfca940ee7